src/bin/inx-chronicle/api/stardust/poi/merkle_hasher.rs

```diff
@@ -1,54 +1,133 @@
 // Copyright 2022 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use std::marker::PhantomData;
+use chronicle::types::stardust::block::BlockId;
+use crypto::hashes::{blake2b::Blake2b256, Digest, Output};
 
-use crypto::hashes::{Digest, Output};
+use super::{
+    merkle_proof::{Hashable, MerkleProof},
+    CreateProofError,
+};
 
 const LEAF_HASH_PREFIX: u8 = 0;
 const NODE_HASH_PREFIX: u8 = 1;
 
-/// A Merkle tree hasher that is generic over the hash function `H` being used.
-pub struct MerkleHasher<H> {
-    _phantom: PhantomData<H>,
-}
-
-impl<H: Default + Digest> MerkleHasher<H> {
-    pub fn new() -> Self {
-        Self { _phantom: PhantomData }
-    }
+/// A Merkle tree hasher.
+pub struct MerkleHasher;
 
-    pub fn hash(&self, data: &[impl AsRef<[u8]>]) -> Output<H> {
+impl MerkleHasher {
+    /// Hash data using the provided hasher type.
+    pub fn hash(data: &[impl AsRef<[u8]>]) -> Output<Blake2b256> {
         match data {
-            [] => self.hash_empty(),
-            [leaf] => self.hash_leaf(leaf),
+            [] => Self::hash_empty(),
+            [leaf] => Self::hash_leaf(leaf),
             _ => {
                 let k = largest_power_of_two(data.len());
-                let l = self.hash(&data[..k]);
-                let r = self.hash(&data[k..]);
-                self.hash_node(&l, &r)
+                let l = Self::hash(&data[..k]);
+                let r = Self::hash(&data[k..]);
+                Self::hash_node(l, r)
             }
         }
     }
 
-    pub fn hash_empty(&self) -> Output<H> {
-        H::digest([])
+    fn hash_empty() -> Output<Blake2b256> {
+        Blake2b256::digest([])
     }
 
-    pub fn hash_leaf(&self, l: impl AsRef<[u8]>) -> Output<H> {
-        let mut hasher = H::default();
+    /// Hash a terminating leaf of the tree.
+    pub fn hash_leaf(l: impl AsRef<[u8]>) -> Output<Blake2b256> {
+        let mut hasher = Blake2b256::default();
         hasher.update([LEAF_HASH_PREFIX]);
         hasher.update(l);
         hasher.finalize()
     }
 
-    pub fn hash_node(&self, l: impl AsRef<[u8]>, r: impl AsRef<[u8]>) -> Output<H> {
-        let mut hasher = H::default();
+    /// Hash a subtree.
+    pub fn hash_node(l: impl AsRef<[u8]>, r: impl AsRef<[u8]>) -> Output<Blake2b256> {
+        let mut hasher = Blake2b256::default();
         hasher.update([NODE_HASH_PREFIX]);
         hasher.update(l);
         hasher.update(r);
         hasher.finalize()
     }
+
```

Comment:
I want everything `proof` related in one file together with the corresponding types. IMO that keeps things better organized.

Reply:
I think it makes more sense to define impls for types where that type is defined.

Reply:
In general yes, but not here. Everything that deals with `proofs` should be in once place IMO so that for the reader everything relevant is nicely placed together. If we cannot agree on that then I'll probably create a new type `MerkleProver` or something to keep "hashing" and "proving" logically separated (proving builds on top of hashing).

Reply:
In that case the hasher should be a module and the stuff you want grouped should be submodules.